### PR TITLE
Macro

### PR DESCRIPTION
**Pull request by rebornix**
Yay! We love PRs! 🎊

Please include a description of your change and ensure:

- [x] Commit message has a short title & issue references
- [x] Each commit does a logical chunk of work.
- [x] It builds and tests pass (e.g `gulp`)

Instead of tracking keystrokes, we track the translation of keystrokes, which are actions or document changes.

### CODE REVIEW

File: src/register/register.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ export enum RegisterMode {
 };
 
 export interface IRegisterContent {
-  text               : string | string[];
+  text               : string | string[] | RecordedState;
```
Comment (johnfn) [inline src/register/register.ts]: Shouldn't it be `RecordedState[]`?

File: src/register/register.ts [hunk 2]
```diff
@@ -70,7 +77,8 @@ export class Register {
     };
   }
 
-  public static putByKey(content: string | string[], register = '"', registerMode = RegisterMode.FigureItOutFromCurrentMode): void {
+  public static putByKey(content: string | string[] | RecordedState, register = '"',
```
Comment (johnfn) [inline src/register/register.ts]: We should probably do `type RegisterContent = string | string[] | RecordedState` at this point.

File: src/register/register.ts (hunk 1 above)
Comment (rebornix) [inline src/register/register.ts]: I directly append each action to `RecordedState.runActions` so we don't necessarily need an array.

File: src/register/register.ts (hunk 2 above)
Comment (rebornix) [inline src/register/register.ts]: Good point

File: src/actions/actions.ts [hunk 3]
```diff
@@ -611,6 +611,14 @@ class CommandInsertRegisterContent extends BaseCommand {
 
     if (register.text instanceof Array) {
        text = (register.text as string []).join("\n");
+    } else if (register.text instanceof RecordedState) {
+      vimState.recordedState.transformations.push({
```
Comment (johnfn) [inline src/actions/actions.ts]: Can we just immediately produce the stringified macro right here instead of shipping it off to some text transformation thing? The transformations are supposed to only be when you're actually changing the content in the text editor.

Comment (johnfn) [inline src/actions/actions.ts]: If we're actually performing a macro at this point, then we absolutely should use a transformation.

But it looked to me like we were just storing the content of the macro in a register, which doesn't require any text transformations.

Comment (rebornix) [inline src/actions/actions.ts]: hmn, I agree. In this particular case, we just need to insert all the keystrokes.

Comment (rebornix) [inline src/actions/actions.ts]: @johnfn just played with Vim again and found that running `Ctrl-R {register}` is actually performing the macro stored in `{register}`.

For example, you recored `ab<BS>c<Esc>` in register `a`, which represents `Enter Insert Mode, insert b, backspace, insert c`. Repeating this macro in Normal mode will just insert `c` and return back to Normal Mode. If you are in Insert Mode and run `<C-r> a`, what gets inserted is `ac`. The first `a` is translated to `Insert a as it's Insert Mode` and then perform all following keystrokes.

**Comment by rebornix** [comment]
Change the code a little bit and fix several corner cases. @johnfn @jpoon @xconverge I think it's good to review then have it merged as I'm not a frequent Macro user so I'd like to have it released first and test in prod :)

**Comment by xconverge** [comment]
This is awesome!

File: test/testUtils.ts [hunk 4]
```diff
@@ -84,4 +84,8 @@ export async function cleanUpWorkspace(): Promise<any> {
 export function setTextEditorOptions(tabSize: number, insertSpaces: boolean): void {
   Configuration.getInstance().tabstop = tabSize;
   Configuration.getInstance().expandtab = insertSpaces;
+  let options = vscode.window.activeTextEditor.options;
+  options.tabSize = tabSize;
```
Comment (rebornix) [inline test/testUtils.ts]: This should be moved once we can really update Code's configuration through API, which relies on #913. So for now we still have to update the cache option in active text editor.
